themis/spec/secure-cell.md

```diff
@@ -11,3 +11,217 @@ Please come back later.
+Sometimes it is acceptable, in other cases you don't have that flexibility.
+
+Seal and Token Protect are basically the same,
+but Token Protect mode stores the authentication tag and other auxiliary data in a detached buffer.
+This allows to replace the original message with an encrypted one of the same length,
+if you can afford to store the auxiliary data elsewhere.
+
+Context Imprint mode exists for the use cases which do not allow for any additional storage.
+This constraint somewhat lowers the security of the cryptosystem
+since there is no space for authentication data or – more importantly – random IV.
+To compensate for this, Secure Cell requires *associated data* to be used in Context Imprint mode.
+
+### Associated data and nonces
+
+Another axis of Secure Cell is how much *associated data* the user provides for encryption.
+
+**Associated data** is used in encryption algorithm, influencing its output,
+but it is not a part of the plaintext, the ciphertext, or the encryption key.
+
+Associated data provides another layer of protection against unintended disclosure, use, and modification of encrypted data.
+Different associated data makes the same plaintext with the same key to be encrypted into a different ciphertext.
+This makes it much harder for the attacker to guess the content of encrypted messages.
+It also makes it harder to reuse encrypted data verbatim in different contexts (replay attacks).
+
+Some parts of the associated data can be transmitted together with the ciphertext, but some can be omitted.
+This complicates unintended decryption: even if the attacker has obtained the encryption key somehow,
+they still need to get access to the associated data which might be stored elsewhere (like the user's brain).
+
+A related concept is **nonce** – an arbitrary random number which must be used only once in cryptographic communication.
+The AES-GCM and AES-CTR algorithms used by Secure Cell use an *initialisation vector* (IV) as a nonce.
+Since they are effectively streaming ciphers, it is **critical** for security to never reuse nonces:
+the most powerful known attack on AES-GCM is based on IV reuse.
+Secure Cell includes several mitigations for it.
+
+With Secure Cell, in addition to the message and the key, users can provide extra **context data**
+which is used to derive associated data for encryption.
+If the context data is not provided explicitly, Secure Cell derives some from the message length.
+It is also used in key derivation to minimise the risks of key reuse and XOR attacks as well.
+See the [Encryption](#encryption) section for the details.
+
+**Seal** and **Token Protect** modes require additional storage.
+In these modes a completely new, random IV is generated for each encrypted piece of data.
+Thus, user-provided context is optional in Seal and Token Protect modes
+but it can still be used to enhance security.
+
+On the other hand, **Context Imprint** mode does not have any additional storage available.
```

Comment:
> does not have any additional storage available

Sounds confusing (at least, for me). What about `does not require any additional storage`?
Or maybe `use` instead of `require`.

Reply:
> Sounds confusing (at least, for me).

Point taken. I'll reword it as you have suggested.